[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59333 [PyTorch] RFC: optimize c10::optional<ArrayRef<T>>**

Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

swolchok committed:
[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>

Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok commented:
If and when this is committed, we should leave a breadcrumb somewhere indicating that this is a thing that will regress when we upgrade to std::optional, and we'll need to port this optimization over.

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

ezyang commented:
Easy breadcrumb is add a static assert on `c10::optional<ArrayRef<T>>` size with a comment saying what to do when the static assert fails.

ezyang commented on c10/util/ArrayRef.h:
```diff
@@ -54,6 +54,12 @@ class ArrayRef final {
   size_type Length;
 
+  void debugCheckNullptrInvariant() {
+    TORCH_INTERNAL_ASSERT(
```
`DEBUG_ONLY` assert?

ezyang commented on c10/util/Optional.cpp:
```diff
@@ -11,4 +12,10 @@ static_assert(
+    "c10::optional<IntArrayRef> should be trivially copyable");
+static_assert(
+    sizeof(c10::optional<c10::IntArrayRef>) == sizeof(c10::IntArrayRef),
+    "c10::optional<IntArrayRef> should be size-optimized");
```
This is assert is good, just shouldn't be in this file (because it's likely to get deleted when we go to std::optional)

swolchok replied:
if people are going to delete this file without looking at it, what's going to stop them from deleting the contents of Optional.h or Optional_test.cpp without looking? (and which one of those two did you want this moved to)

ezyang replied:
Blugh, IDK, some random test file in ATen/test maybe. Or we can just trust in code review, I guess.

ezyang commented on c10/util/Optional.h:
```diff
@@ -392,28 +408,116 @@ struct trivially_copyable_optimization_optional_base {
+};
+
+// HACK: Optimization for ArrayRef<T>. We take advantage of an unused
+// bit pattern in ArrayRef (inspirted by Arthur O'Dwyer's
```
nit: inspired

ezyang reviewed (approved):
okey dokey

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
debug assert, fix comment, copy static_assert to optional_test.cpp too on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
rebase, try not to trigger -Wclass-memaccess on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
just disable -Wclass-memaccess on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: optimize c10::optional<ArrayRef<T>>"


Code comment should explain this in sufficient detail. In brief, making it 16 bytes should get it to be passed in registers.

Differential Revision: [D28843027](https://our.internmc.facebook.com/intern/diff/D28843027/)

[ghstack-poisoned]
